README.md
# pino-datadog-transport

[![NPM version](https://img.shields.io/npm/v/pino-datadog-transport.svg?style=flat-square)](https://www.npmjs.com/package/pino-datadog-transport)
[![CircleCI](https://circleci.com/gh/theogravity/node-pino-datadog-transport.svg?style=svg)](https://circleci.com/gh/theogravity/pino-datadog-transport)
![built with typescript](https://camo.githubusercontent.com/92e9f7b1209bab9e3e9cd8cdf62f072a624da461/68747470733a2f2f666c61742e62616467656e2e6e65742f62616467652f4275696c74253230576974682f547970655363726970742f626c7565)
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)

A [pino v7+](https://github.com/pinojs/pino) transport for sending logs to [Datadog](https://datadog.com/).

It uses [datadog-api-client-typescript](https://github.com/DataDog/datadog-api-client-typescript) to
send logs using the client [v2.LogsApi#submitLog](https://datadoghq.dev/datadog-api-client-typescript/classes/v2.LogsApi.html) method.

- Performs batch sending of logs on a periodic basis, or when log capacity is reached in overall log batch size or count.
- Will retry failed sends.
- Can disable batch sending and always send for each log entry.

## Table of Contents

<!-- TOC -->

- [Installation](#installation)
- [Configuration options](#configuration-options)
- [Implementing the `onError()` / `onDebug()` callback](#implementing-the-onerror--ondebug-callback)
- [Sending logs to Datadog with pino-socket instead](#sending-logs-to-datadog-with-pino-socket-instead)

<!-- TOC END -->

## Installation

`yarn add pino-datadog-transport`

```typescript
import { LoggerOptions, pino } from 'pino';

const pinoConf: LoggerOptions = {
  level: 'trace',
}

const logger = pino(pinoConf, pino.transport({
  target: 'pino-datadog-transport',
  options: {
    ddClientConf: {
      authMethods: {
        apiKeyAuth: <your datadog API key>
      }
    },
  }
}))
```

## Configuration options

```typescript
export interface DDTransportOptions {
  /**
   * DataDog client configuration parameters.
   * @see https://datadoghq.dev/datadog-api-client-typescript/interfaces/client.Configuration.html
   */
  ddClientConf: ConfigurationParameters
  /**
   * Datadog server config for the client. Use this to change the Datadog server region.
   * @see https://github.com/DataDog/datadog-api-client-typescript/blob/1e1097c68a437894b482701ecbe3d61522429319/packages/datadog-api-client-common/servers.ts#L90
   */
  ddServerConf?: {
    /**
     * The datadog server to use. Default is datadoghq.com.
     * Other values could be:
     * - us3.datadoghq.com
     * - us5.datadoghq.com
     * - datadoghq.eu
     * - ddog-gov.com
     */
    site?: string
    subdomain?: string
    protocol?: string
  }
  /**
   * The integration name associated with your log: the technology from which
   * the log originated. When it matches an integration name, Datadog
   * automatically installs the corresponding parsers and facets.
   * @see https://docs.datadoghq.com/logs/log_collection/?tab=host#reserved-attributes
   */
  ddsource?: string
  /**
   * Comma separated tags associated with your logs. Ex: "env:prod,org:finance"
   */
  ddtags?: string
  /**
   * The name of the application or service generating the log events.
   * It is used to switch from Logs to APM, so make sure you define the same
   * value when you use both products.
   * @see https://docs.datadoghq.com/logs/log_collection/?tab=host#reserved-attributes
   */
  service?: string
  /**
   * Called when the plugin is ready to process logs.
   */
  onInit?: () => void
  /**
   * Error handler for when the submitLog() call fails. See readme on how to
   * properly implement this callback.
   */
  onError?: (err: any, logs?: Array<Record<string, any>>) => void
  /**
   * Define this callback to get debug messages from this transport
   */
  onDebug?: (msg: string) => void
  /**
   * Number of times to retry sending the log before onError() is called.
   * Default is 5.
   */
  retries?: number
  /**
   * Interval in which logs are sent to Datadog.
   * Default is 3000 milliseconds.
   */
  sendIntervalMs?: number
  /**
   * Set to true to disable batch sending and send each log as it comes in. This disables
   * the send interval.
   */
  sendImmediate?: boolean
}
```

## Implementing the `onError()` / `onDebug()` callback

You cannot specify the callbacks directly as they are [not serializable](https://github.com/pinojs/pino-pretty#handling-non-serializable-options).

Doing so will result in the following error:

```
DOMException [DataCloneError]: (e)=>{
} could not be cloned.
```

Instead, you need to create another file that implements it:

```typescript
// Your logger file
const p = pino({}, pino.transport({
    target: join(__dirname, 'pino-datadog-logger.js'),
    options: {
      ddClientConf: {
        authMethods: {
          apiKeyAuth: <your datadog API key>
        }
      },
    }
  }))
```

```typescript
/* eslint-disable */
// pino-datadog-logger.js
// https://github.com/pinojs/pino-pretty#handling-non-serializable-options
// Functions as options on the pino transport config are not serializable as they
// are workers, so we create this worker file which includes our callbacks

module.exports = (opts) => {
  return require('pino-datadog-logger')({
    ...opts,
    onError: (data, logItems) => {
      // Your error handling here
    }
  })
}
```

**Note: Log entries can only be a maximum of 1MB in size. This is a Datadog imposed limit.
This library will call onError() if a log entry is 0.95MB (to account for
serialization and metadata).**

## Sending logs to Datadog with pino-socket instead

It is possible to send logs to Datadog using [raw TCP](https://docs.datadoghq.com/logs/log_collection/?tab=tcp) instead of HTTPS. Datadog
recommends HTTPS as the logs are compress-able, whereas TCP-sent logs are not.
HTTPS also has content length limits, whereas TCP does not.

Datadog recommends sending logs over HTTPS instead of raw TCP, as the latest
version of the Datadog agent uses HTTPS with a TCP fallback:

```
// https://docs.datadoghq.com/agent/logs/log_transport?tab=https
For Agent v6.19+/v7.19+, the default transport used for your logs is
compressed HTTPS instead of TCP for the previous versions.
When the Agent starts, if log collection is enabled, it runs a
HTTPS connectivity test. If successful, then the Agent uses
the compressed HTTPS transport, otherwise the Agent falls back to a TCP transport.
```

However, you can send logs using raw TCP + TLS using [`pino-socket`](https://github.com/pinojs/pino-socket).

See instructions [here](https://github.com/pinojs/pino/issues/1511#issuecomment-1207472871) on how to do this.


CHANGELOG.md
## 1.3.1 / 1.3.2 - Wed Nov 01 2023 22:55:38

(Had CI issues that published two versions but they are the same)

**Contributor:** Daniel Hochman

- fix `setServerVariables is deprecated` warning (#18)

## 1.3.0 - Wed Feb 22 2023 15:00:36

**Contributor:** ooga

- Fix level mapping (#13)

## 1.2.2 - Sun Aug 21 2022 20:19:45

**Contributor:** Theo Gravity

- Documentation updates (#9)

## 1.2.1 - Sun Aug 21 2022 09:23:30

**Contributor:** Theo Gravity

- Add Datadog region support and more debug messages (#8)

- A new config option called `ddServerConf` has been added to configure the server region
- Added `onInit` callback
- Added more debugging messages

## 1.1.4 - Fri Aug 19 2022 00:19:40

**Contributor:** Theo Gravity

- Attempt to fix memory leak (#7)

This changes the internal implementation of how we store logs for batches.

## 1.1.3 - Mon Aug 08 2022 00:35:00

**Contributor:** Theo Gravity

- Fix repeat sends (#4)

## 1.1.2 - Mon Aug 08 2022 00:04:42

**Contributor:** Theo Gravity

- Fix queue empty behavior (#3)

## 1.1.1 - Sun Aug 07 2022 22:19:48

**Contributor:** Theo Gravity

- Add batch sending, retry on send failure (#2)

## 1.0.5 - Sun Aug 07 2022 09:20:55

**Contributor:** Theo Gravity

- Update readme

## 1.0.4 - Sun Aug 07 2022 08:57:14

**Contributor:** Theo Gravity

- Fix CI



src/log-storage.ts
import { HTTPLogItem } from '@datadog/datadog-api-client/dist/packages/datadog-api-client-v2/models/HTTPLogItem'

function generateCurrentPointer() {
  return Date.now() + '_' + Math.floor(Math.random() * 1000)
}

export class LogStorage {
  currentBucket: string
  private logBucket: Record<
    string,
    {
      size: number
      logItems: Array<HTTPLogItem>
    }
  >

  constructor() {
    this.logBucket = {}
    this.newLogBucket()
  }

  private newLogBucket() {
    if (this.currentBucket) {
      delete this.logBucket[this.currentBucket]
    }

    this.currentBucket = generateCurrentPointer()
    this.logBucket[this.currentBucket] = {
      size: 0,
      logItems: [],
    }
  }

  getLogBucketByteSize() {
    return this.logBucket[this.currentBucket].size
  }

  getLogCount() {
    return this.logBucket[this.currentBucket].logItems.length
  }

  finishLogBatch() {
    const logs = this.logBucket[this.currentBucket].logItems

    this.newLogBucket()

    return logs
  }

  addLog(log: HTTPLogItem, logByteSize: number) {
    this.logBucket[this.currentBucket].size += logByteSize
    this.logBucket[this.currentBucket].logItems.push(log)
  }
}


.circleci/deploy.sh
#!/bin/bash

if [[ "${CIRCLE_BRANCH}" == "master" ]]
then
  set -e

  echo "Raising package version and updating CHANGELOG.md"

  git config --global push.default simple
  git config --global user.email "theo@suteki.nu"
  git config --global user.name "CircleCI Publisher"

  # Stash any prior changes to prevent merge conflicts
  git add .
  git stash

  # Make sure to get the latest master (if there were any prior commits)
  git pull origin master

  # Re-apply the stash
  # If there is nothing to apply on the stash, a non-zero exit code happens
  # we pipe an empty echo to prevent this
  git stash apply || echo ""

  # Version bump package.json, stamp CHANGELOG.md
  npm run prepare-publish

  # Changelog is now stamped with the version / time info - add to git
  git add CHANGELOG.md
  git add package.json
  git add package-lock.json

  # Amend the version commit with a ci skip so when we push the commits from the CI
  # The CI does not end up recursively building it

  # This gets the last commit log message
  PKG_VERSION=`npm run --silent version-bump show-version`

  # Appending [skip ci] to the log message
  # Note: --amend does not trigger the pre-commit hooks
  git commit -m "${PKG_VERSION} [skip ci]"

  git tag v${PKG_VERSION}

  # Push the commits back to master and assign a versioned release tag
  # Had to add --force because the pull was getting rejected each time
  git push && git push origin "v${PKG_VERSION}"

  # Publish the package to npm
  echo "Publishing package"
  npm publish
else
  echo "Skipping - branch is not master"
fi


src/process-logs.ts
import { HTTPLogItem } from '@datadog/datadog-api-client/dist/packages/datadog-api-client-v2/models/HTTPLogItem'
import { LogsApiSubmitLogRequest } from '@datadog/datadog-api-client/dist/packages/datadog-api-client-v2/apis/LogsApi'
import { ConfigurationParameters } from '@datadog/datadog-api-client/dist/packages/datadog-api-client-common/configuration'
import { v2 } from '@datadog/datadog-api-client'
import pRetry from 'p-retry'
import exitHook from 'exit-hook'
import { LogStorage } from './log-storage'

// Define log sending limits
// https://docs.datadoghq.com/api/latest/logs/#send-logs

// Use 4.9 MB instead of 5 MB to determine when to stop batching
const LOGS_PAYLOAD_SIZE_LIMIT = 5138022

// Use 0.95 MB instead of 1 MB to account for tags and other metadata
const LOG_SIZE_LIMIT = 996147

const FORCE_SEND_MS = 3000

const MAX_LOG_ITEMS = 995

export interface DDTransportOptions {
  /**
   * DataDog client configuration parameters.
   * @see https://datadoghq.dev/datadog-api-client-typescript/interfaces/client.Configuration.html
   */
  ddClientConf: ConfigurationParameters
  /**
   * Datadog server config for the client. Use this to change the Datadog server region.
   * @see https://github.com/DataDog/datadog-api-client-typescript/blob/1e1097c68a437894b482701ecbe3d61522429319/packages/datadog-api-client-common/servers.ts#L90
   */
  ddServerConf?: {
    /**
     * The datadog server to use. Default is datadoghq.com.
     * Other values could be:
     * - us3.datadoghq.com
     * - us5.datadoghq.com
     * - datadoghq.eu
     * - ddog-gov.com
     */
    site?: string
    subdomain?: string
    protocol?: string
  }
  /**
   * The integration name associated with your log: the technology from which
   * the log originated. When it matches an integration name, Datadog
   * automatically installs the corresponding parsers and facets.
   * @see https://docs.datadoghq.com/logs/log_collection/?tab=host#reserved-attributes
   */
  ddsource?: string
  /**
   * Comma separated tags associated with your logs. Ex: "env:prod,org:finance"
   */
  ddtags?: string
  /**
   * The name of the application or service generating the log events.
   * It is used to switch from Logs to APM, so make sure you define the same
   * value when you use both products.
   * @see https://docs.datadoghq.com/logs/log_collection/?tab=host#reserved-attributes
   */
  service?: string
  /**
   * Called when the plugin is ready to process logs.
   */
  onInit?: () => void
  /**
   * Error handler for when the submitLog() call fails. See readme on how to
   * properly implement this callback.
   */
  onError?: (err: any, logs?: Array<Record<string, any>>) => void
  /**
   * Define this callback to get debug messages from this transport
   */
  onDebug?: (msg: string) => void
  /**
   * Number of times to retry sending the log before onError() is called.
   * Default is 5.
   */
  retries?: number
  /**
   * Interval in which logs are sent to Datadog.
   * Default is 3000 milliseconds.
   */
  sendIntervalMs?: number
  /**
   * Set to true to disable batch sending and send each log as it comes in. This disables
   * the send interval.
   */
  sendImmediate?: boolean
}

export const convertLevel = (level: number | string): string => {
  if (typeof level === 'string') {
    return level
  }

  if (level >= 60) {
    return 'fatal'
  }
  if (level >= 50) {
    return 'error'
  }
  if (level >= 40) {
    return 'warning'
  }
  if (level >= 30) {
    return 'info'
  }
  if (level >= 20) {
    return 'debug'
  }

  return 'trace'
}

interface SendLogOpts {
  apiInstance: v2.LogsApi
  logsToSend: Array<HTTPLogItem>
  bucketName: string
  options: DDTransportOptions
}

function sendLogs({ apiInstance, logsToSend, bucketName, options }: SendLogOpts) {
  pRetry(
    async () => {
      const params: LogsApiSubmitLogRequest = {
        body: logsToSend,
        contentEncoding: 'gzip',
      }

      const result = await apiInstance.submitLog(params)

      if (options.onDebug) {
        options.onDebug(`(${bucketName}) Sending ${logsToSend.length} logs to Datadog completed`)
      }

      return result
    },
    { retries: options.retries ?? 5 },
  ).catch((err) => {
    if (options.onError) {
      options.onError(err, logsToSend)
    }
  })
}

export function processLogBuilder(options: DDTransportOptions, apiInstance: v2.LogsApi) {
  const logStorage = new LogStorage()
  let timer = null

  if (!options.sendImmediate) {
    if (options.onDebug) {
      options.onDebug(`Configured to send logs every ${options.sendIntervalMs || FORCE_SEND_MS}ms`)
    }

    timer = setInterval(() => {
      const logCount = logStorage.getLogCount()
      const currentBucket = logStorage.currentBucket

      if (logCount > 0) {
        if (options.onDebug) {
          options.onDebug(`(${currentBucket}) Sending ${logCount} logs to Datadog on timer`)
        }

        // ...logItems is so if we clear logItems in another run, we won't lose these logs
        sendLogs({
          apiInstance: apiInstance,
          logsToSend: logStorage.finishLogBatch(),
          bucketName: currentBucket,
          options: options,
        })
      }
    }, options.sendIntervalMs || FORCE_SEND_MS)
  }

  if (options.onDebug) {
    options.onDebug('Configuring exit hook')
  }

  exitHook(() => {
    if (logStorage.getLogCount() > 0) {
      // Attempt to send logs on an exit call
      if (options.onDebug) {
        options.onDebug(`Shutdown detected. Attempting to send remaining logs to Datadog`)
      }

      if (timer) {
        clearInterval(timer)
      }

      const currentBucket = logStorage.currentBucket

      sendLogs({
        apiInstance: apiInstance,
        logsToSend: logStorage.finishLogBatch(),
        bucketName: currentBucket,
        options: options,
      })
    }
  })

  if (options.onInit) {
    options.onInit()
  }

  return async function processLogs(source) {
    for await (const obj of source) {
      if (!obj) {
        if (options.onDebug) {
          options.onDebug('Log source object is empty')
        }

        return
      }

      // Datadog uses the date field for timestamp
      obj.date = obj.time
      delete obj.time

      const logItem: HTTPLogItem = {
        message: JSON.stringify({
          ...obj,
          level: convertLevel(obj.level),
        }),
      }

      if (options.ddsource) {
        logItem.ddsource = options.ddsource
      }

      if (options.ddtags) {
        logItem.ddtags = options.ddtags
      }

      if (options.service) {
        logItem.service = options.service
      }

      if (obj.hostname) {
        logItem.hostname = obj.hostname
      }

      const logEntryLength =
        logItem.message.length +
        (logItem?.ddsource?.length || 0) +
        (logItem?.ddtags?.length || 0) +
        (logItem?.hostname?.length || 0) +
        (logItem?.service?.length || 0)

      if (logEntryLength > LOG_SIZE_LIMIT) {
        if (options.onError) {
          options.onError(new Error(`Log entry exceeds size limit of ${LOG_SIZE_LIMIT} bytes: ${logEntryLength}`), [
            logItem,
          ])
        }
      }

      if (options.sendImmediate) {
        if (options.onDebug) {
          options.onDebug(`(send-immediate) Sending log to Datadog`)
        }

        // Don't go through the batch logger if sendImmediate is enabled
        sendLogs({
          apiInstance: apiInstance,
          logsToSend: [logItem],
          bucketName: 'send-immediate',
          options: options,
        })

        continue
      }

      logStorage.addLog(logItem, logEntryLength)

      const logCount = logStorage.getLogCount()
      const shouldSend = logStorage.getLogBucketByteSize() > LOGS_PAYLOAD_SIZE_LIMIT || logCount > MAX_LOG_ITEMS

      if (shouldSend) {
        // Get the bucket name into a variable as finishLogBatch() generates a new one
        const currentBucket = logStorage.currentBucket

        if (options.onDebug) {
          options.onDebug(`(${currentBucket}) Sending ${logCount} logs to Datadog`)
        }

        sendLogs({
          apiInstance: apiInstance,
          logsToSend: logStorage.finishLogBatch(),
          bucketName: currentBucket,
          options: options,
        })
      }
    }
  }
}
